GoogleApiWrappers/src/Google.Apis.Storage.v1.IntegrationTests/UploadObjectTest.cs

```diff
@@ -0,0 +1,122 @@
+using Google.Apis.Upload;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Object = Google.Apis.Storage.v1.Data.Object;
+
+namespace Google.Apis.Storage.v1.IntegrationTests
+{
+    public class UploadObjectTest
+    {
+        private static readonly CloudConfiguration s_config = CloudConfiguration.Instance;
+
+        private static readonly string s_bucket = s_config.TempBucketPrefix + "2";
+
+        [Fact]
+        public void SimpleUpload()
+        {
+            var name = GenerateName();
+            var contentType = "application/octet-stream";
+            var source = GenerateData(100);
+            var result = s_config.Client.UploadObject(s_bucket, name, contentType, source);
+            Assert.Equal(s_bucket, result.Bucket);
+            Assert.Equal(name, result.Name);
+            Assert.Equal(contentType, result.ContentType);
+            ValidateData(source, name);            
+        }
+
+        [Fact]
+        public void UploadWithObject()
+        {
+            var destination = new Object
+            {
+                Bucket = s_bucket,
+                Name = GenerateName(),
+                ContentType = "test/type",
+                ContentDisposition = "attachment",
+                Metadata = new Dictionary<string, string> { { "x", "y" } }
+            };
+            var source = GenerateData(100);
+            var result = s_config.Client.UploadObject(destination, source);
+            Assert.NotSame(destination, result);
+            Assert.Equal(destination.Name, result.Name);
+            Assert.Equal(destination.Bucket, result.Bucket);
+            Assert.Equal(destination.ContentType, result.ContentType);
+            Assert.Equal(destination.ContentDisposition, result.ContentDisposition);
+            Assert.Equal(destination.Metadata, result.Metadata);
+            ValidateData(source, destination.Name);
+        }
+
+        [Fact]
+        public async Task UploadAsyncWithProgress()
+        {
+            var chunks = 2;
+            var name = GenerateName();
+            var contentType = "application/octet-stream";
+            var source = GenerateData(UploadObjectOptions.MinimumChunkSize * chunks);
+            int progressCount = 0;
+            var progress = new Progress<IUploadProgress>(p => progressCount++);
+            var result = await s_config.Client.UploadObjectAsync(s_bucket, name, contentType, source,
+                new UploadObjectOptions { ChunkSize = UploadObjectOptions.MinimumChunkSize },
+                CancellationToken.None, progress);
+            Assert.Equal(chunks + 1, progressCount); // Should start with a 0 progress
+            Assert.Equal(name, result.Name); // Assume the rest of the properties are okay...
+            ValidateData(source, name);
+        }
+
+        [Fact]
+        public void ReplaceObject()
```

Comment:
Based on experience, if you specify the Crc32c/MD5 (not sure about etag), it is validated when you upload, i.e. it should match the uploaded data. I found that out by *not* clearing them in the Object in an earlier version of the test. Now sometimes, that's useful - if I've calculated them myself and want to get the server to validate that the data was correctly received. At other times, it's not useful :(